sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala

```diff
@@ -0,0 +1,641 @@
+    if (value < Character.MIN_CODE_POINT || value > Character.MAX_CODE_POINT) {
+      throw new RegexUnsupportedException(s"Invalid hex digit: $hexDigit")
+    }
+
+    RegexHexDigit(hexDigit)
+  }
+
+  private def isOctalDigit(ch: Char): Boolean = ch >= '0' && ch <= '7'
+
+  private def parseOctalDigit: RegexOctalChar = {
+    // \0n   The character with octal value 0n (0 <= n <= 7)
+    // \0nn  The character with octal value 0nn (0 <= n <= 7)
+    // \0mnn The character with octal value 0mnn (0 <= m <= 3, 0 <= n <= 7)
+
+    def parseOctalDigits(n: Integer): RegexOctalChar = {
+      val octal = pattern.substring(pos, pos + n)
+      pos += n
+      RegexOctalChar(octal)
+    }
+
+    if (!eof() && isOctalDigit(pattern.charAt(pos))) {
+      if (pos + 1 < pattern.length && isOctalDigit(pattern.charAt(pos + 1))) {
+        if (pos + 2 < pattern.length && isOctalDigit(pattern.charAt(pos + 2))
+            && pattern.charAt(pos) <= '3') {
+          parseOctalDigits(3)
+        } else {
+          parseOctalDigits(2)
+        }
+      } else {
+        parseOctalDigits(1)
+      }
+    } else {
+      throw new RegexUnsupportedException(
+        "Invalid octal digit", Some(pos))
+    }
+  }
+
+  /** Determine if we are at the end of the input */
+  private def eof(): Boolean = pos == pattern.length
+
+  /** Advance the index by one */
+  private def skip(): Unit = {
+    if (eof()) {
+      throw new RegexUnsupportedException("Unexpected EOF", Some(pos))
+    }
+    pos += 1
+  }
+
+  /** Get the next character and advance the index by one */
+  private def consume(): Char = {
+    if (eof()) {
+      throw new RegexUnsupportedException("Unexpected EOF", Some(pos))
+    } else {
+      pos += 1
+      pattern.charAt(pos - 1)
+    }
+  }
+
+  /** Consume the next character if it is the one we expect */
+  private def consumeExpected(expected: Char): Char = {
+    val consumed = consume()
+    if (consumed != expected) {
+      throw new RegexUnsupportedException(
+        s"Expected '$expected' but found '$consumed'", Some(pos-1))
+    }
+    consumed
+  }
+
+  /** Peek at the next character without consuming it */
+  private def peek(): Option[Char] = {
+    if (eof()) {
+      None
+    } else {
+      Some(pattern.charAt(pos))
+    }
+  }
+
+  private def consumeInt(): Option[Int] = {
+    val start = pos
+    while (!eof() && peek().exists(_.isDigit)) {
+      skip()
+    }
+    if (start == pos) {
+      None
+    } else {
+      Some(pattern.substring(start, pos).toInt)
+    }
+  }
+
+}
+
+/**
+ * Transpile Java/Spark regular expression to a format that cuDF supports, or throw an exception
+ * if this is not possible.
+ */
+class CudfRegexTranspiler {
+
+  val nothingToRepeat = "nothing to repeat"
+
+  def transpile(pattern: String): String = {
+    // parse the source regular expression
+    val regex = new RegexParser(pattern).parse()
+    // validate that the regex is supported by cuDF
+    validate(regex)
```

Comment:
nit: Do we want to try and validate the size/complexity of the regular expression? I don't know exactly what CUDF does to figure out if it needs a small, medium, large, or crazy big stack/memory, but it looks like we could do something similar, and fall back to the CPU if it is too large.  The main reason for this is because we just dropped the default for spark.rapids.memory.gpu.reserve from 1GiB to 256 MiB. The reason we set it at 1GiB was because of hard coded regular expressions that we used. If we are going to fully support arbitrary regular expressions it would be nice to try and tie these two together in some way so we fall back to the CPU if there is not enough reserved memory, or we let users opt into larger regular expressions, but in the instructions we tell them that they need to increase the reserved memory accordingly.

Again this would probably be better as follow on work.

Reply:
That sounds like a great idea. I will file a follow-on issue for this.

Reply:
Filed as https://github.com/NVIDIA/spark-rapids/issues/4061